src/maybe.ts

```diff
@@ -93,4 +95,8 @@ export default class Maybe<Val> {
   public orLazy(fn: () => Maybe<Val>): Maybe<Val> {
     return this.isNothing() ? fn() : this;
   }
+
+  public filter(fn: (val: Val) => boolean): Maybe<Val> {
+    return fn(this.val) ? this : Maybe.empty;
```

Comment:
I would prefer to see a `this.isNothing()` check here like other similar implementations.